### PR TITLE
ENH: Add click extension types

### PR DESCRIPTION
**Pull request by llllllllll**
Adds click param types for config files.

### CODE REVIEW

File: straitlets/ext/click.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    want to check the schema of the file at read time you should use a
+    :class:`~straitlets.StrictSerializable`.
+    """
+    name = 'YAML-FILE'
+
+    def _reader(self, f):
+        return self.config.from_yaml(f)
+
+
+class Base64ConfigFile(_ConfigFile):
```
Comment (ssanderson) [inline straitlets/ext/click.py]: I probably wouldn't even bother with this one.  I can't think of a circumstance where I'd want a base64 config file.  That's like using the headers of syntatically-valid HTTP requests as a config format.

File: straitlets/ext/click.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        super(_ConfigFile, self).__init__(
+            mode='r',
+            encoding=None,
+            errors='strict',
+            lazy=False,
+            atomic=False,
+        )
+        self.config = config_type
+
+    def _reader(self, f):  # pragma: no cover
```
Comment (ssanderson) [inline straitlets/ext/click.py]: I'd probably just call this `read`.

File: straitlets/ext/click.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    """
+    name = 'JSON-FILE'
+
+    def _reader(self, f):
+        return self.config.from_json(f.read())
+
+
+class YamlConfigFile(_ConfigFile):
+    """A click parameter type for reading a :class:`~straitlets.Serializable`
+    object out of base64 encoded json file. This is often used with stdin input
```
Comment (ssanderson) [inline straitlets/ext/click.py]: wrong doc here?

File: straitlets/tests/test_serializable.py [hunk 4]
```diff
@@ -473,3 +479,28 @@ class Strict(StrictSerializable):
 
     assert Strict(x=1).x == 1
     assert Strict(x=1).to_dict() == {"x": 1}
+
+
+class MultipleErrorsStrict(StrictSerializable):
+    """This is not defined in in the body of a function because it breaks
```
Comment (ssanderson) [inline straitlets/tests/test_serializable.py]: "in in"

File: straitlets/ext/tests/test_click.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    ),
+)
+
+single_error_output = re.compile(
+    dedent(
+        """\
+        ^Usage: main \[OPTIONS\]
+
+        Error: Invalid value for "--config": Failed to validate the schema:
+          No default value found for int trait of <.+?>
```
Comment (ssanderson) [inline straitlets/ext/tests/test_click.py]: It seems unfortunate that the multiple error case provides a name here, but the single error case doesn't.

Comment (ssanderson) [inline straitlets/ext/tests/test_click.py]: Oh, though I guess it's already in the message, which makes me wonder if the multiple error formatting is redundant...

Comment (ssanderson) [inline straitlets/ext/tests/test_click.py]: In any case, I think we should be consistent about how we do the formatting.

File: straitlets/ext/click.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
+from __future__ import absolute_import
+
+from operator import itemgetter
+
+import click
```
Comment (ssanderson) [inline straitlets/ext/click.py]: Should we do any special handling here if the user doesn't have `click` installed?

Comment (llllllllll) [inline straitlets/ext/click.py]: I think that if I write:

```python
from straitlets.ext.click import JsonConfigFile
```
and get `ImportError: No module named 'click'` I will understand what is wrong.

File: straitlets/ext/tests/test_click.py (hunk 5 above)
Comment (llllllllll) [inline straitlets/ext/tests/test_click.py]: The reason I prefix the error string with the name is that a user could have a custom validation function which raises a TraitError but does not include the name of the trait in the message so I wanted it to be more clear when multiple exceptions got raised.

**Comment by llllllllll** [comment]
updated

**Pull request merged**